MINIFICPP-2074 Fix time-period/integer validated properties during lo…

martinzink opened the pull request:
…adConfigureFile

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

szaszm commented on minifi_main/MiNiFiMain.cpp:
```diff
@@ -263,6 +263,7 @@ int main(int argc, char **argv) {
     configure->setHome(minifiHome);
     configure->loadConfigureFile(DEFAULT_NIFI_PROPERTIES_FILE);
 
+    configure->commitChanges();
```
Can we skip this commit, and still have the changed/mapped properties appear as part of the manifest?

It feels wrong to write the config right after reading it, and we will do the mapping of the values anyway while reading. I think persisting them in a batch with the first normal persist is enough, and we can rely on in-memory mapping integers to milliseconds.

szaszm commented on libminifi/src/properties/Properties.cpp:
```diff
@@ -87,13 +128,17 @@ void Properties::loadConfigureFile(const std::filesystem::path& configuration_fi
-    properties_[line.getKey()] = {persisted_value, value, false};
+    bool need_to_persist_new_value = false;
+    formatConfigurationProperty(key, persisted_value, value, need_to_persist_new_value);
+    dirty_ |= need_to_persist_new_value;
```
This operator does bitwise OR, not logical OR. It happens to work out fine, due to the memory representation of booleans, but I'd prefer using logical operators with booleans.

martinzink replied:
sure, it will be more readable changed it in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/ee61abd3c1c0efc9f2d460bd691f854f26e9a5aa#diff-f841426e292e70012a81b0fae92f08df7b13120e17b17a92fc360f2f50ec072fR190

szaszm commented on libminifi/src/properties/Properties.cpp:
```diff
@@ -62,6 +63,46 @@ int Properties::getInt(const std::string &key, int default_value) const {
+  ensureIntegerValidatedPropertyHasNoUnit(configuration_property->second, persisted_value, value, need_to_persist_new_value);
+}
+}  // namespace
+
```
Could you extend the comment above `loadConfigureFile` to specify the kind of mapping that's happening here during load, and the motivation? Just to avoid confusing future readers of the code.

martinzink replied:
good idea, I've added an explanatory comment in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/ee61abd3c1c0efc9f2d460bd691f854f26e9a5aa#diff-f841426e292e70012a81b0fae92f08df7b13120e17b17a92fc360f2f50ec072fR157-R158

martinzink commented:
I've also fixed the data-size validated properties similarly to the time-period  validated ones.
I've changed some properties to be time-period validated in [change nifi_flow_engine_alert_period to be TIME_PERIOD validated](https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/a999bf66b8c9bc1e1c127913db8a292d766bd572) [change nifi_flow_engine_event_driven_time_slice to be TIME_PERIOD val…](https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/e2554b5fa6aa5c61615dfb0c53dc187c1b2f9c13) [change nifi_graceful_shutdown_seconds to be TIME_PERIOD validated](https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/9e3a43ab7315a838bd38f9b99ed2b4aedfe3be21), and also fixed the boolean validated ones to be case-insenstive everywhere (it was inconsistent before).

szaszm commented on libminifi/src/utils/StringUtils.cpp:
```diff
@@ -510,4 +510,26 @@ bool StringUtils::matchesSequence(std::string_view str, const std::vector<std::s
+  const char* begin = input.data();
+  char *end;
+  errno = 0;
+  value = std::strtoll(begin, &end, 0);
```
Can we guarantee that strtoull will not throw?

martinzink replied:
I naively just copied this from timeUtils but looking into it strtoull is guranteed to not throw, stoull is the one that throws this only sets the errno.
But using std::from_chars which doesnt rely on a global error variable is probably a better aproach, so I've refactored and added some tests for it in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/471e5dc9ff8473b04d9f801acd4cc9013dbb3db9

arpadboda reviewed (approved):
LGTM at first look

arpadboda commented:
There is a build error in one of the CI jobs:
```
/home/runner/work/nifi-minifi-cpp/nifi-minifi-cpp/extensions/sftp/processors/SFTPProcessorBase.cpp:142:36: error: ‘port_tmp’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
  142 |       common_properties.proxy_port = static_cast<uint16_t>(port_tmp);
```

fgerlits commented on libminifi/include/utils/StringUtils.h:
```diff
@@ -505,7 +505,7 @@ class StringUtils {
   static bool matchesSequence(std::string_view str, const std::vector<std::string>& patterns);
 
- private:
+  static bool splitToUnitAndValue(std::string_view input, std::string& unit, int64_t& value);
```
can we also switch the arguments, please?
```suggestion
  static bool splitToUnitAndValue(std::string_view input, int64_t& value, std::string& unit);
```
that would look more logical to me

martinzink replied:
sure thing, I've changed the name aswell to better reflect the order of the arguments in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002

lordgamez commented on libminifi/include/core/TypedValues.h:
```diff
@@ -110,20 +110,26 @@ class DataSizeValue : public TransformableValue, public state::response::UInt64V
+        {"KB", 1_KiB}, {"MB", 1_MiB}, {"GB", 1_GiB}, {"TB", 1_TiB}, {"PB", 1_PiB},
     };
+    std::transform(unit_str.begin(), unit_str.end(), unit_str.begin(), ::toupper);
+    auto unit_multiplier_it = unit_map.find(unit_str);
+    if (unit_multiplier_it == unit_map.end())
+      return std::nullopt;
+    return unit_multiplier_it->second;
```
This may be a bit shorter:
```
return unit_map.contains(unit_str) ? unit_map[unit_str] : std::nullopt;
```

adamdebreceni commented on libminifi/src/Configuration.cpp:
```diff
@@ -17,135 +17,135 @@
+  {Configuration::nifi_log_alert_url, gsl::make_not_null(core::StandardValidators::get().VALID_VALIDATOR.get())},
+  {Configuration::nifi_log_alert_ssl_context_service, gsl::make_not_null(core::StandardValidators::get().VALID_VALIDATOR.get())},
+  {Configuration::nifi_log_alert_batch_size, gsl::make_not_null(core::StandardValidators::get().VALID_VALIDATOR.get())},
+  {Configuration::nifi_log_alert_flush_period, gsl::make_not_null(core::StandardValidators::get().VALID_VALIDATOR.get())},
```
should we also change this to time period validator?

martinzink commented on extensions/http-curl/tests/HTTPHandlers.h:
```diff
@@ -556,18 +556,18 @@ class HeartbeatHandler : public ServerAwareHandler {
-        for (const auto& property : minifi::Configuration::CONFIGURATION_PROPERTIES) {
-          if (ranges::find(sensitive_props, property.name) != ranges::end(sensitive_props)) {
+        for (const auto& [property_name, property_validator] : minifi::Configuration::CONFIGURATION_PROPERTIES) {
+          if (ranges::find(sensitive_props, property_name) != ranges::end(sensitive_props)) {
```
Good idea, changed this and an other one couple lines down to use contains 
https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002#diff-2c92592d65b3446aa7ec5a38236b3068aefb9163714268c77658c11d17dfe1baR560

martinzink commented on extensions/standard-processors/tests/unit/ConfigurationTests.cpp:
```diff
@@ -53,4 +54,130 @@ TEST_CASE("Configuration can validate values to be assigned to specific properti
+}
+
+
+TEST_CASE("Configuration misconfigured validated properties with in environmental variables") {
```
fixed it in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002#diff-245bd52b73142917446d5f4b4ca2a752e2a7131635f8609b50502849d4159f28R138

martinzink commented on libminifi/include/EventDrivenSchedulingAgent.h:
```diff
@@ -38,11 +38,15 @@ class EventDrivenSchedulingAgent : public ThreadedSchedulingAgent {
+
+    time_slice_ = configuration->get(Configure::nifi_flow_engine_event_driven_time_slice)
+        | utils::flatMap(utils::timeutils::StringToDuration<std::chrono::milliseconds>)
+        | utils::valueOrElse([] { return std::chrono::milliseconds(DEFAULT_TIME_SLICE_MS);});
```
good idea changed it in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002#diff-e8d87fb1392695bb61fc71860b0895a816856bc637825368c05d7f4031212c85R48-R49 https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002#diff-9d30d2ebf0518c2ec672054d681e45cbee06665a0c0eb377a44d0879f018dc53R26

martinzink commented on libminifi/include/core/TypedValues.h:
```diff
@@ -110,20 +110,26 @@ class DataSizeValue : public TransformableValue, public state::response::UInt64V
+        {"K", 1_KB}, {"M", 1_MB}, {"G", 1_GB}, {"T", 1_TB}, {"P", 1_PB},
+        {"KB", 1_KiB}, {"MB", 1_MiB}, {"GB", 1_GiB}, {"TB", 1_TiB}, {"PB", 1_PiB},
     };
+    std::transform(unit_str.begin(), unit_str.end(), unit_str.begin(), ::toupper);
```
good idea, changed these in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002#diff-e2999aedfde011241ca9ddcd8081eec97c6bd96a5b1f6f883aeb2c13985be13fR122

martinzink commented on libminifi/include/core/TypedValues.h:
```diff
@@ -110,20 +110,26 @@ class DataSizeValue : public TransformableValue, public state::response::UInt64V
-  template<typename T, typename std::enable_if<
-      std::is_integral<T>::value>::type* = nullptr>
-  static bool StringToInt(const std::string &input, T &output) {
+  static std::optional<uint64_t> getUnitMultiplier(std::string unit_str) {
```
good idea, changed this to use int64_t in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002#diff-e2999aedfde011241ca9ddcd8081eec97c6bd96a5b1f6f883aeb2c13985be13fR113

martinzink commented on libminifi/include/properties/Properties.h:
```diff
@@ -113,7 +113,7 @@ class Properties {
    * @param fileName path of the configuration file RELATIVE to MINIFI_HOME set by setHome()
    */
-  void loadConfigureFile(const std::filesystem::path& configuration_file);
+  void loadConfigureFile(const std::filesystem::path& configuration_file, const std::string_view prefix = "");
```
removed it in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002#diff-33abd3468567a377d80f73a7183b9c31a4ecf8b3023bca951ca5c5d696e83caeR116

martinzink commented on libminifi/include/utils/TimeUtil.h:
```diff
@@ -201,38 +203,17 @@ std::optional<TargetDuration> cast_to_matching_unit(std::string& unit, const int
+  if (!StringUtils::splitToUnitAndValue(input, unit, value))
     return std::nullopt;
 
+  std::transform(unit.begin(), unit.end(), unit.begin(), ::tolower);
```
good idea, changed it in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002#diff-4a76905d55704437ae0a7a4ee434a73f057c105c97bdf8756f5747b4fbc65e9fR216

martinzink commented on libminifi/src/c2/C2Agent.cpp:
```diff
@@ -180,17 +180,18 @@ void C2Agent::configure(const std::shared_ptr<Configure> &configure, bool reconf
     if (!reconfigure)
       heart_beat_period_ = 3s;
   }
+  logger_->log_debug("Using %u ms as the heartbeat period", heart_beat_period_.count());
```
nice catch, fixed it in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002#diff-6efb1f7ecfeb30f06ecb4dbda25c6ea4b7bde897b7a174253ed1c5331ec94ce7R194

martinzink commented on libminifi/src/properties/Properties.cpp:
```diff
@@ -62,8 +63,100 @@ int Properties::getInt(const std::string &key, int default_value) const {
+std::optional<std::string> ensureTimePeriodValidatedPropertyHasExplicitUnit(const core::PropertyValidator* const validator, std::string& value) {
+  if (validator != core::StandardValidators::get().TIME_PERIOD_VALIDATOR.get())
+    return std::nullopt;
+  if (value.empty() || !std::all_of(value.begin(), value.end(), ::isdigit))
```
nice catch, fixed it in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002#diff-f841426e292e70012a81b0fae92f08df7b13120e17b17a92fc360f2f50ec072fR78

martinzink commented on libminifi/src/properties/Properties.cpp:
```diff
@@ -62,8 +63,100 @@ int Properties::getInt(const std::string &key, int default_value) const {
+      || validator == core::StandardValidators::get().UNSIGNED_LONG_VALIDATOR.get();
+}
+
+std::optional<uint64_t> stringToDataSize(std::string_view input) {
```
it even makes sense to use signed for these, so I've changed it to use int64_t in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002#diff-f841426e292e70012a81b0fae92f08df7b13120e17b17a92fc360f2f50ec072fR100

martinzink commented on libminifi/src/utils/StringUtils.cpp:
```diff
@@ -510,4 +511,20 @@ bool StringUtils::matchesSequence(std::string_view str, const std::vector<std::s
 }
 
+bool StringUtils::splitToUnitAndValue(std::string_view input, std::string& unit, int64_t& value) {
+  const char* begin = input.data();
+  const char* end = begin + input.size();
```
It does seem nicer however it failed to compile on MSVC so I've reverted it back https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/c7d10838758e9b96631e8bbb99d5ad307e17bd8c  https://godbolt.org/z/jvs91f44o

martinzink commented on libminifi/test/unit/StringUtilsTests.cpp:
```diff
@@ -536,4 +536,25 @@ TEST_CASE("StringUtils::matchesSequence works correctly", "[matchesSequence]") {
   REQUIRE(!StringUtils::matchesSequence("xxxabcxxxdefxxx", {"abc", "abc", "def"}));
 }
+
+TEST_CASE("StringUtils::splitToUnitAndValue tests") {
```
I've included these tests (except for the 017 bananas that would parse as 17 "bananas", and "0x17 bananas" would parse as 0 "x17 bananas")

martinzink commented on libminifi/src/properties/Properties.cpp:
```diff
@@ -62,8 +63,100 @@ int Properties::getInt(const std::string &key, int default_value) const {
+    return std::nullopt;
+  if (auto unit_multiplier = core::DataSizeValue::getUnitMultiplier(unit_str)) {
+    return value * *unit_multiplier;
+  }
```
nice catch, fixed them in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/64c5975e3fda0911820465778fa4d73f16d12002#diff-f841426e292e70012a81b0fae92f08df7b13120e17b17a92fc360f2f50ec072fR114-R116

martinzink commented:
> There is a build error in one of the CI jobs:
> 
> ```
> /home/runner/work/nifi-minifi-cpp/nifi-minifi-cpp/extensions/sftp/processors/SFTPProcessorBase.cpp:142:36: error: ‘port_tmp’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
>   142 |       common_properties.proxy_port = static_cast<uint16_t>(port_tmp);
> ```

I am not sure why does this only fail in this PR :thinking: , but fixed the issue nontheless in https://github.com/apache/nifi-minifi-cpp/pull/1543/commits/423558ad876eea3c8fe0c77021ce62135a419e61

martinzink commented:
> Another error popped up, this time in clang-tidy:
> 
> ```
> /home/runner/work/nifi-minifi-cpp/nifi-minifi-cpp/extensions/sftp/processors/../client/SFTPClient.h:20:10: error: 'curl/curl.h' file not found [clang-diagnostic-error]
> #include <curl/curl.h>
>          ^~~~~~~~~~~~~
> /home/runner/work/nifi-minifi-cpp/nifi-minifi-cpp/extensions/sftp/processors/SFTPProcessorBase.cpp:186:133: error: parameter 'connection' is unused [misc-unused-parameters,-warnings-as-errors]
> void SFTPProcessorBase::addConnectionToCache(const SFTPProcessorBase::ConnectionCacheKey& key, std::unique_ptr<utils::SFTPClient>&& connection) {
>                                                                                                                                     ^~~~~~~~~~
>                                                                                                                                      /*connection*/
> ```

This seems false positive to me...
 [Connection is used (it is moved into connections_)](https://github.com/apache/nifi-minifi-cpp/blob/main/extensions/sftp/processors/SFTPProcessorBase.cpp#L203)

fgerlits commented:
> This seems false positive to me... [Connection is used (it is moved into connections_)](https://github.com/apache/nifi-minifi-cpp/blob/main/extensions/sftp/processors/SFTPProcessorBase.cpp#L203)

This sort of false positive tends to happen when `compile_commands.json` is not found or is corrupted.  I have triggered a re-run of the CI job, let's see what happens.

szaszm commented on extensions/standard-processors/tests/unit/ConfigurationTests.cpp:
```diff
@@ -53,4 +54,130 @@ TEST_CASE("Configuration can validate values to be assigned to specific properti
+
+  CHECK(minifi::utils::Environment::setEnvironmentVariable("SOME_VARIABLE", "4000"));
+
+  {
+    std::ofstream properties_file(properties_path);
+    properties_file << "compression.cached.log.max.size=${SOME_VARIABLE}" << std::endl;
+    properties_file << "compression.compressed.log.max.size=3000" << std::endl;
+    properties_file.close();
+  }
```
`std::endl` flushes the stream. I think it's not necessary, but certainly not twice.
```suggestion
    std::ofstream{properties_path}
        << "compression.cached.log.max.size=${SOME_VARIABLE}\n"
        << "compression.compressed.log.max.size=3000\n";
```

szaszm commented on extensions/standard-processors/tests/unit/ConfigurationTests.cpp:
```diff
@@ -53,4 +54,130 @@ TEST_CASE("Configuration can validate values to be assigned to specific properti
+
+  auto properties_path = std::filesystem::temp_directory_path() /  "test.properties";
+
+  {
+    std::ofstream properties_file(properties_path);
+    properties_file << "nifi.c2.agent.heartbeat.period=1min" << std::endl;
+    properties_file << "nifi.administrative.yield.duration=30000" << std::endl;
+    properties_file.close();
+  }
```
```suggestion
  std::ofstream{properties_path}
      << "nifi.c2.agent.heartbeat.period=1min\n"
      << "nifi.administrative.yield.duration=30000\n";
```

szaszm commented on libminifi/include/utils/StringUtils.h:
```diff
@@ -88,7 +87,15 @@ class StringUtils {
   static std::optional<bool> toBool(const std::string& input);
 
-  static std::string toLower(std::string_view str);
+  static inline std::string toLower(std::string_view str) {
+    const auto tolower = [](auto c) { return std::tolower(static_cast<unsigned char>(c)); };
+    return str | ranges::views::transform(tolower) | ranges::to<std::string>();
+  }
+
+  static inline std::string toUpper(std::string_view str)  {
+    const auto toupper = [](auto c) { return std::toupper(static_cast<unsigned char>(c)); };
+    return str | ranges::views::transform(toupper) | ranges::to<std::string>();
+  }
```
The dumber version of taking a `std::string` by value, and doing in-place transform would be more efficient, because it allows NRVO, and better optimization
demonstration: https://godbolt.org/z/q5qWqGbnx
above version: 3 allocations, 1 copy
std::string version: 1 allocation (of the passed-in string)

We need 1 allocation either way, to construct the result, but by taking a std::string by value, we push this responsibility to the caller, who may be able to avoid it by reusing an existing std::string object. If the string is moved in from an existing object, the resulting assembly is basically equivalent to passing in a pointer to the string, and doing an in-place transform, with no extra allocation, but the interface keeps the value-semantics, and falls back to copy when necessary.

```suggestion
  static inline std::string toLower(std::string str) {
    const auto tolower = [](auto c) { return std::tolower(static_cast<unsigned char>(c)); };
    std::transform(std::begin(str), std::end(str), std::begin(str), tolower);
    return str;
  }

  static inline std::string toUpper(std::string str)  {
    const auto toupper = [](auto c) { return std::toupper(static_cast<unsigned char>(c)); };
    std::transform(std::begin(str), std::end(str), std::begin(str), toupper);
    return str;
  }
```

szaszm commented on libminifi/src/Configuration.cpp:
```diff
@@ -179,9 +179,9 @@ std::vector<std::string> Configuration::getSensitiveProperties(const std::functi
-  for (const auto& config_property: Configuration::CONFIGURATION_PROPERTIES) {
-    if (config_property.name == property_name) {
-      return config_property.validator->validate(property_name, property_value).valid();
+  for (const auto& [config_property_name, config_property_validator]: Configuration::CONFIGURATION_PROPERTIES) {
+    if (config_property_name == property_name) {
+      return config_property_validator->validate(property_name, property_value).valid();
```
Now that it's not a vector anymore, there is a more efficient way to do lookup in CONFIGURATION_PROPERTIES

szaszm commented on libminifi/src/Configuration.cpp:
```diff
@@ -179,12 +179,10 @@ std::vector<std::string> Configuration::getSensitiveProperties(const std::functi
-    }
-  }
-  return true;
+  if (!Configuration::CONFIGURATION_PROPERTIES.contains(property_name))
+    return true;
+
+  return Configuration::CONFIGURATION_PROPERTIES.at(property_name)->validate(property_name, property_value).valid();
```
We should do the lookup only once, not twice.
```suggestion
  const auto validator = Configuration::CONFIGURATION_PROPERTIES.find(property_name);
  if (validator == std::end(Configuration::CONFIGURATION_PROPERTIES))
    return true;

  return (*validator)->validate(property_name, property_value).valid();
```

szaszm reviewed (approved):
LGTM, but we need a rebase :)

szaszm commented on libminifi/src/Configuration.cpp:
```diff
@@ -17,140 +17,140 @@
+  {Configuration::nifi_metrics_publisher_metrics, gsl::make_not_null(core::StandardValidators::get().VALID_VALIDATOR.get())},
+  {Configuration::controller_socket_enable, gsl::make_not_null(core::StandardValidators::get().BOOLEAN_VALIDATOR.get())},
+  {Configuration::controller_socket_local_any_interface, gsl::make_not_null(core::StandardValidators::get().BOOLEAN_VALIDATOR.get())},
+  {Configuration::controller_socket_host,gsl::make_not_null(core::StandardValidators::get().VALID_VALIDATOR.get())},
+  {Configuration::controller_socket_port, gsl::make_not_null(core::StandardValidators::get().PORT_VALIDATOR.get())},
+  {Configuration::controller_ssl_context_service,gsl::make_not_null(core::StandardValidators::get().VALID_VALIDATOR.get())}
```
```
/home/szaszm/nifi-minifi-cpp-3/libminifi/src/Configuration.cpp:151:  Missing space after ,  [whitespace/comma] [3]
/home/szaszm/nifi-minifi-cpp-3/libminifi/src/Configuration.cpp:153:  Missing space after ,  [whitespace/comma] [3]
```
